homeassistant/components/huawei_lte/config_flow.py

```diff
@@ -0,0 +1,204 @@
+                try:
+                    conn = AuthorizedConnection(
+                        user_input[CONF_URL], username="", password=""
+                    )
+                    user_input[CONF_USERNAME] = ""
+                    user_input[CONF_PASSWORD] = ""
+                except ResponseErrorException:
+                    _LOGGER.debug(
+                        "Could not login with empty credentials, proceeding unauthenticated",
+                        exc_info=True,
+                    )
+                    conn = Connection(user_input[CONF_URL])
+                    del user_input[CONF_USERNAME]
+                    del user_input[CONF_PASSWORD]
+        except LoginErrorUsernameWrongException:
+            errors[CONF_USERNAME] = "incorrect_username"
+        except LoginErrorPasswordWrongException:
+            errors[CONF_PASSWORD] = "incorrect_password"
+        except LoginErrorUsernamePasswordWrongException:
+            errors[CONF_USERNAME] = "incorrect_username_or_password"
+        except LoginErrorUsernamePasswordOverrunException:
+            errors["base"] = "login_attempts_exceeded"
+        except ResponseErrorException:
+            _LOGGER.warning("Response error", exc_info=True)
+            errors["base"] = "response_error"
+        except BaseException:
+            _LOGGER.warning("Connection error", exc_info=True)
+            errors[CONF_URL] = "connection_failed"
+        if errors:
+            logout()
+            return await self._async_show_user_form(
+                user_input=user_input, errors=errors
+            )
+
+        title = None
+        client = Client(conn)
+        try:
+            info = client.device.basic_information()
+            title = info["devicename"]
+        except BaseException:
+            _LOGGER.debug(
+                "Could not get device.basic_information[devicename]", exc_info=True
+            )
+        if not title:
+            try:
+                info = client.device.information()
+                title = info["DeviceName"]
+            except BaseException:
+                _LOGGER.debug(
+                    "Could not get device.information[DeviceName]", exc_info=True
+                )
+        logout()
+
+        return self.async_create_entry(
+            title=title or DEFAULT_DEVICE_NAME, data=user_input
+        )
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Huawei LTE options flow."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry):
+        """Initialize options flow."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Handle options flow."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        fields = [
+            (vol.Required(key, default=self.config_entry.options.get(key, True)), bool)
+            for key in chain(DEVICE_TRACKER_KEYS, SENSOR_KEYS, SWITCH_KEYS)
```

Comment:
I suggest we just set a default disabled value for each entity type and not implement any integration options for this. The user can turn off new entities for this integration in the system options, and or disable individual entities via entity registry.

Reply:
If I understand correctly, this suggestion again goes against the goal of being able to invoke a lot of different requests to the device, even ones that end up being used at all. Every request counts.

Reply:
If the user disables an entity there won't be a request, I'm thinking.

Reply:
Defaulting to disabled sounds to me as being against the spirit of ADR-0003 which kind of says everything should be enabled/exposed by default.

Reply:
We have specifically added system level options to disable entities to allow the user to control this, so we don't need to implement the options on a integration level.

Reply:
So, something like this?

1. At startup, fetch everything once. A fetch is necessary so we know what entities exist in the first place, and to get data for possible new component features that use data from categories that other existing features already use.
1. Create missing entities as default disabled.
1. Stop polling categories for which there are no enabled entities.
1. Start polling categories if one enables an entity that uses info from them.

The above would create a problem with the device tracker as its entities come and go; disabling its fetch doesn't work that well with the above basic rules. So it will probably need some more configuration in order to continue fetching even with all entities disabled, if one wants new ones to appear. ISTR there was something related already implemented in some other integrations, do you remember any?